File: frontend/src/services/ong.ts
/**
 * Service API pour le module ONG — bénéficiaires, programmes, missions.
 */
import { clientAPI } from "./client_api";

export interface BeneficiaireONG {
  id: string;
  ong_id: string;
  nom: string;
  digiid: string | null;
  programme: string;
  zone: string | null;
  date_inscription: string;
  statut: string;
  notes: string | null;
}

export interface ProgrammeONG {
  id: string;
  ong_id: string;
  nom: string;
  description: string | null;
  zone: string | null;
  budget: number | null;
  date_debut: string;
  date_fin: string | null;
  statut: string;
}

export interface MissionTerrain {
  id: string;
  ong_id: string;
  programme_id: string | null;
  titre: string;
  zone: string | null;
  date_depart: string;
  date_retour: string | null;
  objectifs: string | null;
  statut: string;
}

export interface StatsONG {
  nb_beneficiaires: number;
  nb_programmes: number;
  nb_missions: number;
  zones: string[];
}

export async function listerBeneficiaires(): Promise<BeneficiaireONG[]> {
  return clientAPI.get<BeneficiaireONG[]>("/api/v1/utilisateur/ong/beneficiaires", {
    authentifie: true,
  });
}

export async function creerBeneficiaire(data: {
  nom: string;
  digiid?: string;
  programme: string;
  zone?: string;
  notes?: string;
}): Promise<BeneficiaireONG> {
  return clientAPI.post<BeneficiaireONG>("/api/v1/utilisateur/ong/beneficiaires", data, {
    authentifie: true,
  });
}

export async function listerProgrammes(): Promise<ProgrammeONG[]> {
  return clientAPI.get<ProgrammeONG[]>("/api/v1/utilisateur/ong/programmes", {
    authentifie: true,
  });
}

export async function creerProgramme(data: {
  nom: string;
  description?: string;
  zone?: string;
  budget?: number;
  date_debut: string;
  date_fin?: string;
}): Promise<ProgrammeONG> {
  return clientAPI.post<ProgrammeONG>("/api/v1/utilisateur/ong/programmes", data, {
    authentifie: true,
  });
}

export async function listerMissions(): Promise<MissionTerrain[]> {
  return clientAPI.get<MissionTerrain[]>("/api/v1/utilisateur/ong/missions", {
    authentifie: true,
  });
}

export async function creerMission(data: {
  programme_id?: string;
  titre: string;
  zone?: string;
  date_depart: string;
  date_retour?: string;
  objectifs?: string;
}): Promise<MissionTerrain> {
  return clientAPI.post<MissionTerrain>("/api/v1/utilisateur/ong/missions", data, {
    authentifie: true,
  });
}

export async function obtenirStats(): Promise<StatsONG> {
  return clientAPI.get<StatsONG>("/api/v1/utilisateur/ong/stats", {
    authentifie: true,
  });
}

File: backend/src/modeles/ong.py
# -*- coding: utf-8 -*-
"""
Modèles ONG — Bénéficiaires, programmes, missions terrain.
Avec cloisonnement par domaine et département.
"""
import uuid
from datetime import date, datetime, timezone
from sqlalchemy import Column, String, Text, Date, DateTime, ForeignKey, Integer, Float, Boolean
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from src.base_donnees.base import Base


class BeneficiaireONG(Base):
    __tablename__ = "beneficiaires_ong"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    ong_id = Column(UUID(as_uuid=True), ForeignKey("utilisateur.id"), nullable=False, index=True)
    nom = Column(String(255), nullable=False)
    digiid = Column(String(50), nullable=True, index=True)
    programme = Column(String(255), nullable=False)
    zone = Column(String(100), nullable=True)
    date_inscription = Column(DateTime, default=datetime.utcnow, nullable=False)
    statut = Column(String(20), default="actif")
    notes = Column(Text, nullable=True)

    # ✅ AJOUT : Lien vers une mission (optionnel)
    mission_id = Column(
        UUID(as_uuid=True),
        ForeignKey("missions_terrain.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
        comment="Mission associée (optionnel)"
    )

    # --- Cloisonnement ---
    domaine_id = Column(
        UUID(as_uuid=True),
        ForeignKey("domaines.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
        comment="Domaine de rattachement"
    )
    departement_id = Column(
        UUID(as_uuid=True),
        ForeignKey("departements.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
        comment="Département de rattachement"
    )

    ong = relationship("Utilisateur", backref="beneficiaires_ong")
    # ✅ AJOUT : Relation avec MissionTerrain
    mission = relationship("MissionTerrain", backref="beneficiaires", lazy="select")


class ProgrammeONG(Base):
    __tablename__ = "programmes_ong"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    ong_id = Column(UUID(as_uuid=True), ForeignKey("utilisateur.id"), nullable=False, index=True)
    nom = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    zone = Column(String(100), nullable=True)
    budget = Column(Float, nullable=True)
    date_debut = Column(Date, nullable=False)
    date_fin = Column(Date, nullable=True)
    statut = Column(String(20), default="actif")

    # --- Cloisonnement ---
    domaine_id = Column(
        UUID(as_uuid=True),
        ForeignKey("domaines.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
        comment="Domaine de rattachement"
    )
    departement_id = Column(
        UUID(as_uuid=True),
        ForeignKey("departements.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
        comment="Département de rattachement"
    )

    ong = relationship("Utilisateur", backref="programmes_ong")
    # ✅ Relation avec les rapports
    rapports = relationship("Rapport", back_populates="programme", lazy="select")


class MissionTerrain(Base):
    __tablename__ = "missions_terrain"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    ong_id = Column(UUID(as_uuid=True), ForeignKey("utilisateur.id"), nullable=False, index=True)
    programme_id = Column(UUID(as_uuid=True), ForeignKey("programmes_ong.id"), nullable=True)
    titre = Column(String(255), nullable=False)
    zone = Column(String(100), nullable=True)
    date_depart = Column(Date, nullable=False)
    date_retour = Column(Date, nullable=True)
    objectifs = Column(Text, nullable=True)
    statut = Column(String(20), default="planifiee")

    # --- Cloisonnement ---
    domaine_id = Column(
        UUID(as_uuid=True),
        ForeignKey("domaines.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
        comment="Domaine de rattachement"
    )
    departement_id = Column(
        UUID(as_uuid=True),
        ForeignKey("departements.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
        comment="Département de rattachement"
    )

    ong = relationship("Utilisateur", backref="missions_terrain")
    programme = relationship("ProgrammeONG", backref="missions")
    # ✅ Relation avec les rapports
    rapports = relationship("Rapport", back_populates="mission", lazy="select")


class MissionAgent(Base):
    """Table de liaison entre missions et agents."""
    __tablename__ = "mission_agent"
    
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    
    # ✅ CORRECTION : "missions_terrain.id" au lieu de "mission_terrain.id"
    mission_id = Column(
        UUID(as_uuid=True), 
        ForeignKey("missions_terrain.id", ondelete="CASCADE"), 
        nullable=False
    )
    agent_id = Column(
        UUID(as_uuid=True), 
        ForeignKey("utilisateur.id", ondelete="CASCADE"), 
        nullable=False
    )
    
    instructions = Column(Text, nullable=True)
    date_limite = Column(Date, nullable=True)
    statut = Column(String(50), default="assignee")  # assignee, en_cours, terminee, annulee
    date_assignation = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    date_debut = Column(DateTime(timezone=True), nullable=True)
    date_completion = Column(DateTime(timezone=True), nullable=True)
    
    # ✅ Relations explicites
    mission = relationship(
        "MissionTerrain",
        foreign_keys=[mission_id],
        primaryjoin="MissionAgent.mission_id == MissionTerrain.id",
        backref="agents_assignes",
        lazy="selectin",
    )
    agent = relationship(
        "Utilisateur",
        foreign_keys=[agent_id],
        primaryjoin="MissionAgent.agent_id == Utilisateur.id",
        backref="missions_assignees",
        lazy="selectin",
    )